### PR TITLE
--mount option for `create` & `run` command

### PR DESCRIPTION
**Pull request by kunalkushwaha**
Signed-off-by: Kunal Kushwaha <kushwaha_kunal_v7@lab.ntt.co.jp>

### CODE REVIEW

**Comment by kunalkushwaha** [comment]
Fixes #1192 

`tmpfs` support is still required to be added.

If possible, please review if in general this approach is fine. 
- Few logic are common in  `--volume` & `--mount` case. But modifying those functions will be bigger change, so added new functions. Maybe can be optimized in separate PR. 

//cc @rhatdan

File: cmd/podman/create_cli.go [hunk 1]
```diff
@@ -74,6 +74,49 @@ func addWarning(warnings []string, msg string) []string {
 	return append(warnings, msg)
 }
 
+func parseMounts(mounts []string) error {
```
Comment (rhatdan) [inline cmd/podman/create_cli.go]: Can't you combine this function and the function below to create the mounts struct into one function?
 func (c *CreateConfig) GetFSMounts(specMounts []spec.Mount) ([]spec.Mount, error) {

**Comment by rhatdan** [comment]
Missing changes to:
         Man Pages
         Command Completions.
         Need tests

You need to check what happens if I have existing mounts that you are mounting over, those should be removed from the spec file.

For examples

podman run -v /var/lib/mariadb:/var/lib/mariadb --mount type=bind,src=/var/lib/mariadb,dest=/var/lib/mariadb ...

Or better yet

podman run  --mount type=bind,src=/etc/resolv.conf,dest=/etc/resolv.conf ...

Is the following legal syntax?

podman run  --mount type=tmpfs,dest=/dev/shm ...

**Comment by rhatdan** [comment]
@mheon @baude @umohnani8 PTAL

File: pkg/spec/spec.go [hunk 2]
```diff
@@ -306,6 +306,11 @@ func CreateConfigToOCISpec(config *CreateConfig) (*spec.Spec, error) { //nolint
 	if err := config.GetVolumesFrom(); err != nil {
 		return nil, errors.Wrap(err, "error getting volume mounts from --volumes-from flag")
 	}
+	fsMounts, err := config.GetFSMounts(configSpec.Mounts)
```
Comment (mheon) [inline pkg/spec/spec.go]: Do we want this to supercede mounts given by `-v` in precedence? I think we might want it to, in which case it should be moved below `GetVolumeMounts`

Also, we need to handle overlapping mounts already in the spec, as we do for the mounts returned by `GetVolumeMounts`

Comment (kunalkushwaha) [reply pkg/spec/spec.go]: So here I understand you mean, `specs mount`  <  `-v mount` < `--mount`
i.e. any mount point defined in specs is superceded by `-v` mounts.
&    mount point defined by `-v` mount  will be superceded by ``-mount`` option

File: cmd/podman/create_cli.go (hunk 1 above)
Comment (kunalkushwaha) [reply cmd/podman/create_cli.go]: There are little more required to just merge these functions, as [`parseMount()`](https://github.com/containers/libpod/pull/1432/files#diff-958bbd628fbbafabf2cdfc21a7dbe118R80) uses other functions present in `create_cli.go` file, whereas [`GetFSMounts()`](https://github.com/containers/libpod/pull/1432/files#diff-cebb6d9e2dd46ffe9665b195619c5c13R226) has some common logic that is shared with functions of [`/pkg/createconfig.go`](https://github.com/containers/libpod/blob/7b4489a945bd8eddfd791bd5e2e888ee348c9ea5/pkg/spec/createconfig.go#L150) like `GetVolumeMounts()`.

I think, reorganizing code for various mount options (`-v` , `--mount`, `--volume-from`) to some common function will be better, not can be done in separate PR.

**Comment by rhatdan** [comment]
@kunalkushwaha Looks like your tests are failing.

File: cmd/podman/create_cli.go [hunk 3]
```diff
@@ -74,6 +74,70 @@ func addWarning(warnings []string, msg string) []string {
 	return append(warnings, msg)
 }
 
+// Format supported.
+// podman run --mount type=bind,src=/etc/resolv.conf,target=/etc/resolv.conf ...
+// podman run --mount type=tmpfs,target=/dev/shm ..
+func parseMounts(mounts []string) error {
+	errInvalidSyntax := errors.Errorf("incorrect mount format : should be --mount type=<bind|tmpfs>,src=<host-dir>,target=<ctr-dir>,[option]")
```
Comment (TomSweeneyRedHat) [inline cmd/podman/create_cli.go]: options not option no?

Comment (TomSweeneyRedHat) [reply cmd/podman/create_cli.go]: and no commas?  I'd remove them unless they're part of the syntax.

Comment (rhatdan) [reply cmd/podman/create_cli.go]: They are part of the syntax

**Comment by rhatdan** [comment]
I still don't like the fact we are parsing the mount string twice.  Can't you consolodate them some how?

File: cmd/podman/create_cli.go (hunk 3 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/create_cli.go]: oops, scratch the comma remarks, I see it below.

File: docs/podman-run.1.md [hunk 4]
```diff
@@ -653,6 +653,21 @@ Set the UTS mode for the container
 
 **NOTE**: the host mode gives the container access to changing the host's hostname and is therefore considered insecure.
 
+**−−mount**[=*type=<bind|tmpfs>,<src|source>=HOST−DIR,<target|dst,destination>=CONTAINER−DIR[:OPTIONS]*]
+The --mount flag allows you to mount host-directories and tmpfs mounts in a container.
+The --mount flag supports most options that are supported by the -v|--volume flag, but uses a different syntax.
+
+The type must be bind or tmpfs .tmpfs mounted folder data will be lost after container stops.
```
Comment (TomSweeneyRedHat) [inline docs/podman-run.1.md]: Suggest:  The type must be bind or tmpfs.  Note: the .tmpfs mounted directory will not be retained after the container stops.

File: docs/podman-run.1.md [hunk 5]
```diff
@@ -653,6 +653,21 @@ Set the UTS mode for the container
...
+
+The source can be specified as "source" or "src" & target can be specified as "target","dst" or "destination".
+The CONTAINER−DIR must be an absolute path such as /src/docs. The HOST−DIR must be an absolute path as well.
+podman bind −mounts the HOST−DIR to the path you specify.
```
Comment (TomSweeneyRedHat) [inline docs/podman-run.1.md]: suggest "you specify" to " that is specified".

File: docs/podman-run.1.md [hunk 6]
```diff
@@ -653,6 +653,21 @@ Set the UTS mode for the container
...
+For example, if you supply the /foo value, podman creates a bind −mount.You can specify multiple −−mount options to mount
+one or more mounts to a container.
+
+You can add :ro or :rw suffix to a volume to mount it read −only or read −write mode, respectively. By default, the volumes are mounted read −write. See examples.
+
```
Comment (TomSweeneyRedHat) [inline docs/podman-run.1.md]: I'm not a fan of "you" in man pages.  Can you please change that?  "For example if /foo is supplied as the value, Podman create a bind --mount.  Multiple --mount options can be specified to mount one or more mounts to a container.

The :ro or :rw suffix can be added to a volume ..." 

etc.

**Comment by kunalkushwaha** [comment]
> I still don't like the fact we are parsing the mount string twice. Can't you consolodate them some how?

It can done by storing the [`Mounts`](https://github.com/containers/libpod/pull/1432/files#diff-cebb6d9e2dd46ffe9665b195619c5c13R127) information in  `spec.Mount` struct instead of `string`. I used string to just keep the consistency with `Volumes` & `VolumesFrom` elements.
``` patch

+	Mounts             []string              //mounts
	Volumes            []string              //volume
	VolumesFrom        []string
```

**Comment by kunalkushwaha** [comment]
@rhatdan I am still there. Will update it now.

**Comment by kunalkushwaha** [comment]
I think still only _travis-ci_ is only invoked, not others.

File: pkg/spec/createconfig.go [hunk 7]
```diff
@@ -221,6 +222,53 @@ func (c *CreateConfig) GetVolumeMounts(specMounts []spec.Mount) ([]spec.Mount, e
...
+				return nil, errors.Wrapf(err, "relabel failed %q", mount.Source)
+			}
+		}
+		if foundZ {
+			if err := label.Relabel(mount.Source, c.MountLabel, false); err != nil {
+				return nil, errors.Wrapf(err, "relabel failed %q", mount.Source)
+			}
+		}
+		if rootProp == "" {
+			mount.Options = append(mount.Options, "private")
```
Comment (rhatdan) [inline pkg/spec/createconfig.go]: Should be rprivate

**Comment by rhatdan** [comment]
papr must be down.

File: pkg/spec/createconfig.go (hunk 7 above)
Comment (kunalkushwaha) [reply pkg/spec/createconfig.go]: Sorry for that. Fixed now.

**Comment by rhatdan** [comment]
@mheon @baude @umohnani8 @giuseppe @vrothberg  PTAL

**Comment by rhatdan** [comment]
@kunalkushwaha Could you rebase,  I would love to get this in this week.

**Comment by kunalkushwaha** [comment]
Sorry it won't b possible to me to work before Oct 1st as travelling and
don't have stable internet and laptop to work.

Please feel free to update it is it's possible.

Regards,
Kunal Kushwaha

------------------------------

**Comment by rhatdan** [comment]
Ok @kunalkushwaha I will take it from here.  Thanks for your contribution.

**Comment by rhatdan** [comment]
Just take a look at all the issues, and pick out one not assigned, or if it is assigned, ping the person assigned and see if he is actually working on it.
Thanks.

**Pull request closed without merging**